sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/display/DisplayData.java

```diff
@@ -508,55 +258,67 @@ public String getLabel() {
     @JsonGetter("linkUrl")
     @JsonInclude(JsonInclude.Include.NON_NULL)
     @Nullable
-    public String getLinkUrl() {
-      return url;
-    }
+    public abstract String getLinkUrl();
 
-    @Override
-    public String toString() {
-      return String.format("%s:%s=%s", ns, key, value);
+    public static <T> Item<T> create(String key, Type type, @Nullable T value) {
+      FormattedItemValue formatted = safeFormat(type, value);
+      return of(null, key, type, formatted.getLongValue(), formatted.getShortValue(), null, null);
     }
 
-    @Override
-    public boolean equals(Object obj) {
-      if (obj instanceof Item) {
-        Item that = (Item) obj;
-        return Objects.equals(this.ns, that.ns)
-            && Objects.equals(this.key, that.key)
-            && Objects.equals(this.type, that.type)
-            && Objects.equals(this.value, that.value)
-            && Objects.equals(this.shortValue, that.shortValue)
-            && Objects.equals(this.label, that.label)
-            && Objects.equals(this.url, that.url);
+    private static FormattedItemValue safeFormat(Type type, @Nullable Object value) {
+      if (value == null) {
+        return FormattedItemValue.DEFAULT;
       }
 
-      return false;
+      return type.format(value);
     }
 
-    @Override
-    public int hashCode() {
-      return Objects.hash(
-          this.ns,
-          this.key,
-          this.type,
-          this.value,
-          this.shortValue,
-          this.label,
-          this.url);
+    public Item<T> withNamespace(Class<?> namespace) {
+      checkNotNull(namespace);
+      return withNamespace(ClassForDisplay.of(namespace));
     }
 
-    private Item withLabel(String label) {
-      return new Item(this.ns, this.key, this.type, this.value, this.shortValue, this.url, label);
+    private Item<T> withNamespace(ClassForDisplay namespace) {
+      checkNotNull(namespace);
+      return withNamespace(namespaceOf(namespace));
     }
 
-    private Item withUrl(String url) {
-      return new Item(this.ns, this.key, this.type, this.value, this.shortValue, url, this.label);
+    public Item<T> withNamespace(String namespace) {
+      checkNotNull(namespace);
+      return of(
+          namespace, getKey(), getType(), getValue(), getShortValue(), getLabel(), getLinkUrl());
     }
 
-    private Item withNamespace(ClassForDisplay nsClass) {
-      String namespace = namespaceOf(nsClass);
-      return new Item(
-          namespace, this.key, this.type, this.value, this.shortValue, this.url, this.label);
+    public Item<T> withLabel(String label) {
```

Comment:
The current behavior is to accept and ignore null values for label/url. This could be useful:

* To clear a previously-set label. We don't really have a scenario for this, but there's no reason you shouldn't be able to clear a label/link for an item.
* To support consumers registering computed values that may be null at runtime. We could also provide a family of `with<Property>IfNotNull(value)` functions. We did this for the builder APIs so that errors in registering display data in the normal APIs would be fail-fast. I see now that having consistency in this API set would be desirable.

I'm leaning towards flipping the current behavior (fail on null/empty), though we have zero real usage of these APIs to validate which behavior is most desirable. We could also remove them altogether.

I've added javadoc comments describing the current behavior.